{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "source": [
    "### Running DESeq2\n",
    "\n",
    "As I went looking for single-cell differential expression packages, I came across [this paper](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-019-2599-6) and unabashedly skipped to the conclusion which reported that \"methods developed specifically for scRNAseq data do not show significantly better performance compared to the methods designed for bulk RNAseq data; and methods that consider behavior of each individual gene (not all genes) in calling DE genes outperform the other tools.\"\n",
    "\n",
    "Luckily, I have already used DESeq2 (which is designed for bulk-seq data). It also _is_ one of the methods that considers the behavior of each individual gene (according to the same paper). For those reasons, I'm applying it here. \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"DESeq2\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define a path prefix:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "prefix <- \"/data/codec/production.run/mrna/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define the directory with the counts and then a results directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "countsdir <- paste(prefix,\"de.csvs/\",sep=\"\")\n",
    "resdir <- paste(prefix,\"deseq2.res/\",sep=\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "List the files in `countsdir`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "sampleFiles <- list.files(countsdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'A.col.csv'</li><li>'A.cts.csv'</li><li>'B.col.csv'</li><li>'B.cts.csv'</li><li>'G.col.csv'</li><li>'G.cts.csv'</li><li>'P.col.csv'</li><li>'P.cts.csv'</li><li>'R.col.csv'</li><li>'R.cts.csv'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'A.col.csv'\n",
       "\\item 'A.cts.csv'\n",
       "\\item 'B.col.csv'\n",
       "\\item 'B.cts.csv'\n",
       "\\item 'G.col.csv'\n",
       "\\item 'G.cts.csv'\n",
       "\\item 'P.col.csv'\n",
       "\\item 'P.cts.csv'\n",
       "\\item 'R.col.csv'\n",
       "\\item 'R.cts.csv'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'A.col.csv'\n",
       "2. 'A.cts.csv'\n",
       "3. 'B.col.csv'\n",
       "4. 'B.cts.csv'\n",
       "5. 'G.col.csv'\n",
       "6. 'G.cts.csv'\n",
       "7. 'P.col.csv'\n",
       "8. 'P.cts.csv'\n",
       "9. 'R.col.csv'\n",
       "10. 'R.cts.csv'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"A.col.csv\" \"A.cts.csv\" \"B.col.csv\" \"B.cts.csv\" \"G.col.csv\" \"G.cts.csv\"\n",
       " [7] \"P.col.csv\" \"P.cts.csv\" \"R.col.csv\" \"R.cts.csv\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sampleFiles"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "File names are returned in alphabetical order, so the `col.csv` always comes before the `cts.csv`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "converting counts to integer mode\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates: 30 workers\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "-- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "final dispersion estimates, fitting model and testing: 30 workers\n",
      "\n",
      "-- replacing outliers and refitting for 27 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates: 30 workers\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "-- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "final dispersion estimates, fitting model and testing: 30 workers\n",
      "\n",
      "-- replacing outliers and refitting for 24 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates: 30 workers\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "-- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "final dispersion estimates, fitting model and testing: 30 workers\n",
      "\n",
      "-- replacing outliers and refitting for 30 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates: 30 workers\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "-- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "final dispersion estimates, fitting model and testing: 30 workers\n",
      "\n",
      "-- replacing outliers and refitting for 35 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates: 30 workers\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "-- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "final dispersion estimates, fitting model and testing: 30 workers\n",
      "\n",
      "-- replacing outliers and refitting for 27 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates: 30 workers\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "-- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "final dispersion estimates, fitting model and testing: 30 workers\n",
      "\n",
      "-- replacing outliers and refitting for 31 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in DESeqDataSet(se, design = design, ignoreRank): design has a single variable, with all samples having the same value.\n  use instead a design of '~ 1'. estimateSizeFactors, rlog and the VST can then be used\n",
     "output_type": "error",
     "traceback": [
      "Error in DESeqDataSet(se, design = design, ignoreRank): design has a single variable, with all samples having the same value.\n  use instead a design of '~ 1'. estimateSizeFactors, rlog and the VST can then be used\nTraceback:\n",
      "1. DESeqDataSetFromMatrix(countData = subct, colData = subcoldata, \n .     design = ~COND)",
      "2. DESeqDataSet(se, design = design, ignoreRank)",
      "3. stop(\"design has a single variable, with all samples having the same value.\\n  use instead a design of '~ 1'. estimateSizeFactors, rlog and the VST can then be used\")"
     ]
    }
   ],
   "source": [
    "for (cond in c(\"A\",\"B\",\"G\",\"P\",\"R\")) {\n",
    "    \n",
    "    cond_files <- grep(cond, sampleFiles, value = TRUE)\n",
    "    \n",
    "    cts <- as.matrix(read.csv(paste(countsdir, \n",
    "                                    cond_files[2], # using the second position, which is the cts.csv\n",
    "                                    sep=\"\"\n",
    "                                   ),row.names=1, check.names = FALSE))\n",
    "    \n",
    "    coldata <- read.csv(paste(countsdir, \n",
    "                              cond_files[1], # using the first position, which is the col.csv\n",
    "                              sep=\"\"\n",
    "                             ), row.names=1)\n",
    "    \n",
    "    # reclassify as factor, right now reading at an integer\n",
    "    coldata$FID <- as.factor(coldata$FID)\n",
    "    \n",
    "# #     # first, run the time course model outlined in the rnaseqgene vignette\n",
    "# #     ddsTC <- DESeqDataSetFromMatrix(countData = cts,\n",
    "# #                     colData = coldata,\n",
    "# #                                 design = ~ cond + TIME + cond:TIME\n",
    "# #                                )\n",
    "# #     ddsTC <- DESeq(ddsTC, test=\"LRT\", reduced = ~ cond + TIME)\n",
    "# #     res <- results(ddsTC)\n",
    "# #     write.csv(as.data.frame(res), file=paste(resdir,cond,\".TC.csv\",sep=\"\"))\n",
    "    \n",
    "    # then for each time point, do just regular differential expression between condulation and control\n",
    "    for (celltype in c(\"B_Memory\", \"B_Naive\", \"CD4_T_Memory\", \"CD4_T_Naive\", \n",
    "                       \"CD8_T_Memory_MAIT_GD\", \"CD8_T_Naive\", \"HSC\", \"NK\", \n",
    "                       \"pDC\", \"cDC\", \"Mono_cDC_All\")) {\n",
    "        subct <- subset(cts, select=grep(paste(as.character(celltype),\"-\",sep=\"\"),colnames(cts),value = TRUE))\n",
    "        subcoldata <- subset(coldata, CT == celltype)\n",
    "        dds <- DESeqDataSetFromMatrix(countData = subct,\n",
    "                              colData = subcoldata,\n",
    "                              design = ~ COND\n",
    "                             )\n",
    "        dds <- DESeq(dds, parallel = TRUE)\n",
    "        res <- results(dds, contrast = c(\"COND\",cond,\"C\"))\n",
    "        write.csv(as.data.frame(res), file=paste(resdir, cond,\".\",as.character(celltype),\".csv\",sep=\"\"))\n",
    "    }\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "converting counts to integer mode\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates: 30 workers\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "-- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "final dispersion estimates, fitting model and testing: 30 workers\n",
      "\n",
      "-- replacing outliers and refitting for 12 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates: 30 workers\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "-- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "final dispersion estimates, fitting model and testing: 30 workers\n",
      "\n",
      "-- replacing outliers and refitting for 22 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates: 30 workers\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "-- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "final dispersion estimates, fitting model and testing: 30 workers\n",
      "\n",
      "-- replacing outliers and refitting for 7 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates: 30 workers\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "-- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "final dispersion estimates, fitting model and testing: 30 workers\n",
      "\n",
      "-- replacing outliers and refitting for 31 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates: 30 workers\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "-- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "final dispersion estimates, fitting model and testing: 30 workers\n",
      "\n",
      "-- replacing outliers and refitting for 15 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates: 30 workers\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "-- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "final dispersion estimates, fitting model and testing: 30 workers\n",
      "\n",
      "-- replacing outliers and refitting for 29 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates: 30 workers\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "-- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "final dispersion estimates, fitting model and testing: 30 workers\n",
      "\n",
      "-- replacing outliers and refitting for 19 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "fitting model and testing\n",
      "\n",
      "converting counts to integer mode\n",
      "\n",
      "estimating size factors\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "gene-wise dispersion estimates: 30 workers\n",
      "\n",
      "mean-dispersion relationship\n",
      "\n",
      "-- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "final dispersion estimates, fitting model and testing: 30 workers\n",
      "\n",
      "-- replacing outliers and refitting for 13 genes\n",
      "-- DESeq argument 'minReplicatesForReplace' = 7 \n",
      "-- original counts are preserved in counts(dds)\n",
      "\n",
      "estimating dispersions\n",
      "\n",
      "fitting model and testing\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for (cond in c(\"A\",\"B\",\"G\",\"R\")) {\n",
    "    \n",
    "    cond_files <- grep(cond, sampleFiles, value = TRUE)\n",
    "    \n",
    "    cts <- as.matrix(read.csv(paste(countsdir, \n",
    "                                    cond_files[2], # using the second position, which is the cts.csv\n",
    "                                    sep=\"\"\n",
    "                                   ),row.names=1, check.names = FALSE))\n",
    "    \n",
    "    coldata <- read.csv(paste(countsdir, \n",
    "                              cond_files[1], # using the first position, which is the col.csv\n",
    "                              sep=\"\"\n",
    "                             ), row.names=1)\n",
    "    \n",
    "    # reclassify as factor, right now reading at an integer\n",
    "    coldata$FID <- as.factor(coldata$FID)\n",
    "    \n",
    "# #     # first, run the time course model outlined in the rnaseqgene vignette\n",
    "# #     ddsTC <- DESeqDataSetFromMatrix(countData = cts,\n",
    "# #                     colData = coldata,\n",
    "# #                                 design = ~ cond + TIME + cond:TIME\n",
    "# #                                )\n",
    "# #     ddsTC <- DESeq(ddsTC, test=\"LRT\", reduced = ~ cond + TIME)\n",
    "# #     res <- results(ddsTC)\n",
    "# #     write.csv(as.data.frame(res), file=paste(resdir,cond,\".TC.csv\",sep=\"\"))\n",
    "    \n",
    "    # then for each time point, do just regular differential expression between condulation and control\n",
    "    for (celltype in c('Mono_NC', 'Mono_C')) {\n",
    "        subct <- subset(cts, select=grep(paste(as.character(celltype),\"-\",sep=\"\"),colnames(cts),value = TRUE))\n",
    "        subcoldata <- subset(coldata, CT == celltype)\n",
    "        dds <- DESeqDataSetFromMatrix(countData = subct,\n",
    "                              colData = subcoldata,\n",
    "                              design = ~ COND\n",
    "                             )\n",
    "        dds <- DESeq(dds, parallel = TRUE)\n",
    "        res <- results(dds, contrast = c(\"COND\",cond,\"C\"))\n",
    "        write.csv(as.data.frame(res), file=paste(resdir, cond,\".\",as.character(celltype),\".csv\",sep=\"\"))\n",
    "    }\n",
    "    \n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
