{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here I make the final preparations for DESeq2 as was done for the bulk RNA-sequencing data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pickle as pkl\n",
    "import numpy as np\n",
    "from tqdm.notebook import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prefix = '/data/codec/production.run/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "mountpoint = '/data/clue/'\n",
    "prefix_adts = mountpoint + 'prod/adts/'\n",
    "prefix_mrna = mountpoint + 'prod/mrna/'\n",
    "prefix_comb = mountpoint + 'prod/comb/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# `ct2`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## mRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = pd.read_pickle(prefix_mrna + 'pkls/pseudobulk_ct2.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Just get the raw data, that's all I want."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "T8_Naive    384\n",
       "B_Mem       384\n",
       "NK          384\n",
       "pDC         384\n",
       "B_Naive     384\n",
       "HSC         384\n",
       "T4_EM       384\n",
       "M_cDC       384\n",
       "T_Tox       384\n",
       "T4_Naive    384\n",
       "ncM         320\n",
       "cM          320\n",
       "cDC         320\n",
       "Name: CT, dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts['CT'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['B_Mem', 'B_Naive', 'HSC', 'M_cDC', 'NK', 'T4_EM', 'T4_Naive',\n",
       "       'T8_Naive', 'T_Tox', 'cDC', 'cM', 'ncM', 'pDC'], dtype=object)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(counts['CT'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts.isna().sum().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create 5 separate sets of files, one for each stimulation, that has counts for both controls and the stim conditions, and in the appropriate format for easy reading into DESeq2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "223295311703421ea7305538817f4cd2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/5 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for cond in tqdm(['A','B','G','P','R']):\n",
    "    # extract out only the cond conditions and controls\n",
    "    df = counts[(counts['COND'] == cond) | (counts['COND'] == 'C')]\n",
    "    \n",
    "    # make a new index that encompasses the entirety of the sample name\n",
    "    df.index = ['-'.join([i,j,k]) for i,j,k in zip(df['CT'], df['COND'], df['FID'])]\n",
    "    \n",
    "\n",
    "    # With the bulk data, I had to remove and sum over duplicate genes\n",
    "    # but because this is pseuobulked from an adata object, I already ran \n",
    "    # var_names_make_unique, so I think I'm good (I also confirmed in a separate cell)\n",
    "    \n",
    "    # extract out only the genes, then rotate because that's what DESeq2 expects\n",
    "    cts = df.iloc[:,3:].T\n",
    "    \n",
    "    \n",
    "    # get new columns, and then make a separate dfs for the coldata\n",
    "    coldata_columns = df.columns[:3]\n",
    "    coldata = pd.DataFrame(data=df[coldata_columns].values, index=cts.columns, columns=coldata_columns)\n",
    "    \n",
    "    \n",
    "    # next line was required, was getting the following error from DESeq2:\n",
    "    # every gene contains at least one zero, cannot compute log geometric means\n",
    "    # this was the suggested fix:\n",
    "    cts = cts.loc[cts.sum(1) > 0,:] + 1\n",
    "\n",
    "    #export to csv\n",
    "    cts.to_csv(prefix_mrna + 'vals/de/all/ct2/input/%s_cts.csv' % cond)\n",
    "    coldata.to_csv(prefix_mrna + 'vals/de/all/ct2/input/%s_col.csv' % cond)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ADTs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = pd.read_pickle(prefix_adts + 'pkls/pseudobulk_ct2.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CT</th>\n",
       "      <th>COND</th>\n",
       "      <th>FID</th>\n",
       "      <th>CD103|ITGAE</th>\n",
       "      <th>CD137|TNFRSF9</th>\n",
       "      <th>CD274|CD274</th>\n",
       "      <th>CD11b|ITGAM</th>\n",
       "      <th>CD39|ENTPD1</th>\n",
       "      <th>CD197|CCR7</th>\n",
       "      <th>CD20|MS4A1</th>\n",
       "      <th>...</th>\n",
       "      <th>CD49a|ITGA1</th>\n",
       "      <th>CD133|PROM</th>\n",
       "      <th>CD357|TNFRSF18</th>\n",
       "      <th>CD270|TNFRSF14</th>\n",
       "      <th>IL-21R|IL21R</th>\n",
       "      <th>B7-H4|VTCN1</th>\n",
       "      <th>CD26|DPP4</th>\n",
       "      <th>CD155|PVR</th>\n",
       "      <th>CD30|TNFRSF8</th>\n",
       "      <th>CD49E|ITGA5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T4_Naive</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>118.0</td>\n",
       "      <td>1590.0</td>\n",
       "      <td>987.0</td>\n",
       "      <td>805.0</td>\n",
       "      <td>1028.0</td>\n",
       "      <td>1907.0</td>\n",
       "      <td>236.0</td>\n",
       "      <td>...</td>\n",
       "      <td>320.0</td>\n",
       "      <td>313.0</td>\n",
       "      <td>790.0</td>\n",
       "      <td>1637.0</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>2463.0</td>\n",
       "      <td>6362.0</td>\n",
       "      <td>1131.0</td>\n",
       "      <td>668.0</td>\n",
       "      <td>1248.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T4_Naive</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>199.0</td>\n",
       "      <td>1950.0</td>\n",
       "      <td>1438.0</td>\n",
       "      <td>774.0</td>\n",
       "      <td>1329.0</td>\n",
       "      <td>2990.0</td>\n",
       "      <td>379.0</td>\n",
       "      <td>...</td>\n",
       "      <td>436.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>1017.0</td>\n",
       "      <td>1981.0</td>\n",
       "      <td>2645.0</td>\n",
       "      <td>3334.0</td>\n",
       "      <td>5773.0</td>\n",
       "      <td>1577.0</td>\n",
       "      <td>1020.0</td>\n",
       "      <td>1603.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T4_Naive</td>\n",
       "      <td>A</td>\n",
       "      <td>10</td>\n",
       "      <td>501.0</td>\n",
       "      <td>5193.0</td>\n",
       "      <td>3877.0</td>\n",
       "      <td>2439.0</td>\n",
       "      <td>3893.0</td>\n",
       "      <td>6985.0</td>\n",
       "      <td>1121.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1254.0</td>\n",
       "      <td>1210.0</td>\n",
       "      <td>2989.0</td>\n",
       "      <td>5679.0</td>\n",
       "      <td>7913.0</td>\n",
       "      <td>9536.0</td>\n",
       "      <td>19177.0</td>\n",
       "      <td>4672.0</td>\n",
       "      <td>2866.0</td>\n",
       "      <td>3612.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T4_Naive</td>\n",
       "      <td>A</td>\n",
       "      <td>11</td>\n",
       "      <td>139.0</td>\n",
       "      <td>1732.0</td>\n",
       "      <td>1207.0</td>\n",
       "      <td>674.0</td>\n",
       "      <td>1178.0</td>\n",
       "      <td>2347.0</td>\n",
       "      <td>358.0</td>\n",
       "      <td>...</td>\n",
       "      <td>369.0</td>\n",
       "      <td>384.0</td>\n",
       "      <td>955.0</td>\n",
       "      <td>1940.0</td>\n",
       "      <td>2613.0</td>\n",
       "      <td>3226.0</td>\n",
       "      <td>5598.0</td>\n",
       "      <td>1538.0</td>\n",
       "      <td>879.0</td>\n",
       "      <td>1332.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T4_Naive</td>\n",
       "      <td>A</td>\n",
       "      <td>12</td>\n",
       "      <td>203.0</td>\n",
       "      <td>2777.0</td>\n",
       "      <td>1857.0</td>\n",
       "      <td>1052.0</td>\n",
       "      <td>1891.0</td>\n",
       "      <td>3274.0</td>\n",
       "      <td>495.0</td>\n",
       "      <td>...</td>\n",
       "      <td>629.0</td>\n",
       "      <td>599.0</td>\n",
       "      <td>1515.0</td>\n",
       "      <td>2977.0</td>\n",
       "      <td>3999.0</td>\n",
       "      <td>5079.0</td>\n",
       "      <td>7580.0</td>\n",
       "      <td>2146.0</td>\n",
       "      <td>1405.0</td>\n",
       "      <td>2027.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>M_cDC</td>\n",
       "      <td>R</td>\n",
       "      <td>62</td>\n",
       "      <td>134.0</td>\n",
       "      <td>3320.0</td>\n",
       "      <td>2444.0</td>\n",
       "      <td>6729.0</td>\n",
       "      <td>1420.0</td>\n",
       "      <td>1220.0</td>\n",
       "      <td>268.0</td>\n",
       "      <td>...</td>\n",
       "      <td>602.0</td>\n",
       "      <td>324.0</td>\n",
       "      <td>909.0</td>\n",
       "      <td>2307.0</td>\n",
       "      <td>1939.0</td>\n",
       "      <td>2473.0</td>\n",
       "      <td>1786.0</td>\n",
       "      <td>1147.0</td>\n",
       "      <td>763.0</td>\n",
       "      <td>11072.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>M_cDC</td>\n",
       "      <td>R</td>\n",
       "      <td>63</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>M_cDC</td>\n",
       "      <td>R</td>\n",
       "      <td>7</td>\n",
       "      <td>353.0</td>\n",
       "      <td>6702.0</td>\n",
       "      <td>17348.0</td>\n",
       "      <td>12877.0</td>\n",
       "      <td>2107.0</td>\n",
       "      <td>1628.0</td>\n",
       "      <td>290.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1105.0</td>\n",
       "      <td>342.0</td>\n",
       "      <td>1073.0</td>\n",
       "      <td>3566.0</td>\n",
       "      <td>2115.0</td>\n",
       "      <td>2496.0</td>\n",
       "      <td>3555.0</td>\n",
       "      <td>1199.0</td>\n",
       "      <td>1030.0</td>\n",
       "      <td>16470.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>M_cDC</td>\n",
       "      <td>R</td>\n",
       "      <td>8</td>\n",
       "      <td>148.0</td>\n",
       "      <td>2819.0</td>\n",
       "      <td>6210.0</td>\n",
       "      <td>3463.0</td>\n",
       "      <td>1342.0</td>\n",
       "      <td>1460.0</td>\n",
       "      <td>220.0</td>\n",
       "      <td>...</td>\n",
       "      <td>489.0</td>\n",
       "      <td>311.0</td>\n",
       "      <td>861.0</td>\n",
       "      <td>1944.0</td>\n",
       "      <td>1829.0</td>\n",
       "      <td>2745.0</td>\n",
       "      <td>1844.0</td>\n",
       "      <td>1073.0</td>\n",
       "      <td>779.0</td>\n",
       "      <td>8548.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>M_cDC</td>\n",
       "      <td>R</td>\n",
       "      <td>9</td>\n",
       "      <td>678.0</td>\n",
       "      <td>15417.0</td>\n",
       "      <td>37167.0</td>\n",
       "      <td>24157.0</td>\n",
       "      <td>5873.0</td>\n",
       "      <td>5586.0</td>\n",
       "      <td>1144.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2596.0</td>\n",
       "      <td>1252.0</td>\n",
       "      <td>3610.0</td>\n",
       "      <td>9344.0</td>\n",
       "      <td>7920.0</td>\n",
       "      <td>9767.0</td>\n",
       "      <td>7297.0</td>\n",
       "      <td>4626.0</td>\n",
       "      <td>3459.0</td>\n",
       "      <td>46142.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4800 rows × 99 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           CT COND FID  CD103|ITGAE  CD137|TNFRSF9  CD274|CD274  CD11b|ITGAM  \\\n",
       "0    T4_Naive    A   0        118.0         1590.0        987.0        805.0   \n",
       "1    T4_Naive    A   1        199.0         1950.0       1438.0        774.0   \n",
       "2    T4_Naive    A  10        501.0         5193.0       3877.0       2439.0   \n",
       "3    T4_Naive    A  11        139.0         1732.0       1207.0        674.0   \n",
       "4    T4_Naive    A  12        203.0         2777.0       1857.0       1052.0   \n",
       "..        ...  ...  ..          ...            ...          ...          ...   \n",
       "315     M_cDC    R  62        134.0         3320.0       2444.0       6729.0   \n",
       "316     M_cDC    R  63          0.0            0.0          0.0          0.0   \n",
       "317     M_cDC    R   7        353.0         6702.0      17348.0      12877.0   \n",
       "318     M_cDC    R   8        148.0         2819.0       6210.0       3463.0   \n",
       "319     M_cDC    R   9        678.0        15417.0      37167.0      24157.0   \n",
       "\n",
       "     CD39|ENTPD1  CD197|CCR7  CD20|MS4A1  ...  CD49a|ITGA1  CD133|PROM  \\\n",
       "0         1028.0      1907.0       236.0  ...        320.0       313.0   \n",
       "1         1329.0      2990.0       379.0  ...        436.0       450.0   \n",
       "2         3893.0      6985.0      1121.0  ...       1254.0      1210.0   \n",
       "3         1178.0      2347.0       358.0  ...        369.0       384.0   \n",
       "4         1891.0      3274.0       495.0  ...        629.0       599.0   \n",
       "..           ...         ...         ...  ...          ...         ...   \n",
       "315       1420.0      1220.0       268.0  ...        602.0       324.0   \n",
       "316          0.0         0.0         0.0  ...          0.0         0.0   \n",
       "317       2107.0      1628.0       290.0  ...       1105.0       342.0   \n",
       "318       1342.0      1460.0       220.0  ...        489.0       311.0   \n",
       "319       5873.0      5586.0      1144.0  ...       2596.0      1252.0   \n",
       "\n",
       "     CD357|TNFRSF18  CD270|TNFRSF14  IL-21R|IL21R  B7-H4|VTCN1  CD26|DPP4  \\\n",
       "0             790.0          1637.0        2019.0       2463.0     6362.0   \n",
       "1            1017.0          1981.0        2645.0       3334.0     5773.0   \n",
       "2            2989.0          5679.0        7913.0       9536.0    19177.0   \n",
       "3             955.0          1940.0        2613.0       3226.0     5598.0   \n",
       "4            1515.0          2977.0        3999.0       5079.0     7580.0   \n",
       "..              ...             ...           ...          ...        ...   \n",
       "315           909.0          2307.0        1939.0       2473.0     1786.0   \n",
       "316             0.0             0.0           0.0          0.0        0.0   \n",
       "317          1073.0          3566.0        2115.0       2496.0     3555.0   \n",
       "318           861.0          1944.0        1829.0       2745.0     1844.0   \n",
       "319          3610.0          9344.0        7920.0       9767.0     7297.0   \n",
       "\n",
       "     CD155|PVR  CD30|TNFRSF8  CD49E|ITGA5  \n",
       "0       1131.0         668.0       1248.0  \n",
       "1       1577.0        1020.0       1603.0  \n",
       "2       4672.0        2866.0       3612.0  \n",
       "3       1538.0         879.0       1332.0  \n",
       "4       2146.0        1405.0       2027.0  \n",
       "..         ...           ...          ...  \n",
       "315     1147.0         763.0      11072.0  \n",
       "316        0.0           0.0          0.0  \n",
       "317     1199.0        1030.0      16470.0  \n",
       "318     1073.0         779.0       8548.0  \n",
       "319     4626.0        3459.0      46142.0  \n",
       "\n",
       "[4800 rows x 99 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Just get the raw data, that's all I want."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "T8_Naive    384\n",
       "B_Mem       384\n",
       "NK          384\n",
       "pDC         384\n",
       "B_Naive     384\n",
       "HSC         384\n",
       "T4_EM       384\n",
       "M_cDC       384\n",
       "T_Tox       384\n",
       "T4_Naive    384\n",
       "ncM         320\n",
       "cM          320\n",
       "cDC         320\n",
       "Name: CT, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts['CT'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['B_Mem', 'B_Naive', 'HSC', 'M_cDC', 'NK', 'T4_EM', 'T4_Naive',\n",
       "       'T8_Naive', 'T_Tox', 'cDC', 'cM', 'ncM', 'pDC'], dtype=object)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(counts['CT'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts.isna().sum().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create 5 separate sets of files, one for each stimulation, that has counts for both controls and the stim conditions, and in the appropriate format for easy reading into DESeq2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4286da6d184d4a3ea2c46838375c6188",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/5 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for cond in tqdm(['A','B','G','P','R']):\n",
    "    # extract out only the cond conditions and controls\n",
    "    df = counts[(counts['COND'] == cond) | (counts['COND'] == 'C')]\n",
    "    \n",
    "    # make a new index that encompasses the entirety of the sample name\n",
    "    df.index = ['-'.join([i,j,k]) for i,j,k in zip(df['CT'], df['COND'], df['FID'])]\n",
    "    \n",
    "\n",
    "    # With the bulk data, I had to remove and sum over duplicate genes\n",
    "    # but because this is pseuobulked from an adata object, I already ran \n",
    "    # var_names_make_unique, so I think I'm good (also confirmed in a separate cell)\n",
    "    \n",
    "    # extract out only the genes, then rotate because that's what DESeq2 expects\n",
    "    cts = df.iloc[:,3:].T\n",
    "    \n",
    "    \n",
    "    # get new columns, and then make a separate dfs for the coldata\n",
    "    coldata_columns = df.columns[:3]\n",
    "    coldata = pd.DataFrame(data=df[coldata_columns].values, index=cts.columns, columns=coldata_columns)\n",
    "    \n",
    "#     if cond == 'P':\n",
    "#         cts = cts.loc[:,~np.any(np.stack([cts.columns.str.contains('Mono_C-C'), cts.columns.str.contains('Mono_NC-C')]), axis=0)]\n",
    "#         coldata = coldata.loc[~np.any(np.stack([coldata.index.str.contains('Mono_C-C'), coldata.index.str.contains('Mono_NC-C')]), axis=0),:]\n",
    "    \n",
    "    # next line was required, was getting the following error from DESeq2:\n",
    "    # every gene contains at least one zero, cannot compute log geometric means\n",
    "    # this was the suggested fix:\n",
    "    cts = cts.loc[cts.sum(1) > 0,:] + 1\n",
    "\n",
    "    #export to csv\n",
    "    cts.to_csv(prefix_adts + 'vals/de/all/ct2/input/%s_cts.csv' % cond)\n",
    "    coldata.to_csv(prefix_adts + 'vals/de/all/ct2/input/%s_col.csv' % cond)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# `ct3`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## mRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = pd.read_pickle(prefix_mrna + 'pkls/pseudobulk_ct3.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CT</th>\n",
       "      <th>COND</th>\n",
       "      <th>FID</th>\n",
       "      <th>AL627309.1</th>\n",
       "      <th>AL627309.3</th>\n",
       "      <th>AL669831.5</th>\n",
       "      <th>LINC00115</th>\n",
       "      <th>FAM41C</th>\n",
       "      <th>AL645608.3</th>\n",
       "      <th>SAMD11</th>\n",
       "      <th>...</th>\n",
       "      <th>AC145212.1</th>\n",
       "      <th>MAFIP</th>\n",
       "      <th>AC011043.1</th>\n",
       "      <th>AL592183.1</th>\n",
       "      <th>AC007325.4</th>\n",
       "      <th>AL354822.1</th>\n",
       "      <th>AC004556.1</th>\n",
       "      <th>AC233755.2</th>\n",
       "      <th>AC233755.1</th>\n",
       "      <th>AC240274.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MAIT</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MAIT</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MAIT</td>\n",
       "      <td>A</td>\n",
       "      <td>10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MAIT</td>\n",
       "      <td>A</td>\n",
       "      <td>11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MAIT</td>\n",
       "      <td>A</td>\n",
       "      <td>12</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6075</th>\n",
       "      <td>cDC2</td>\n",
       "      <td>R</td>\n",
       "      <td>62</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6076</th>\n",
       "      <td>cDC2</td>\n",
       "      <td>R</td>\n",
       "      <td>63</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6077</th>\n",
       "      <td>cDC2</td>\n",
       "      <td>R</td>\n",
       "      <td>7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6078</th>\n",
       "      <td>cDC2</td>\n",
       "      <td>R</td>\n",
       "      <td>8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6079</th>\n",
       "      <td>cDC2</td>\n",
       "      <td>R</td>\n",
       "      <td>9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6080 rows × 20606 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        CT COND FID  AL627309.1  AL627309.3  AL669831.5  LINC00115  FAM41C  \\\n",
       "0     MAIT    A   0         0.0         0.0         0.0        0.0     0.0   \n",
       "1     MAIT    A   1         0.0         0.0         1.0        3.0     0.0   \n",
       "2     MAIT    A  10         0.0         0.0         1.0        0.0     0.0   \n",
       "3     MAIT    A  11         0.0         0.0         0.0        0.0     0.0   \n",
       "4     MAIT    A  12         0.0         0.0         0.0        0.0     0.0   \n",
       "...    ...  ...  ..         ...         ...         ...        ...     ...   \n",
       "6075  cDC2    R  62         0.0         0.0         0.0        0.0     0.0   \n",
       "6076  cDC2    R  63         0.0         0.0         0.0        0.0     0.0   \n",
       "6077  cDC2    R   7         0.0         0.0         0.0        0.0     0.0   \n",
       "6078  cDC2    R   8         0.0         0.0         3.0        0.0     0.0   \n",
       "6079  cDC2    R   9         0.0         0.0         0.0        1.0     0.0   \n",
       "\n",
       "      AL645608.3  SAMD11  ...  AC145212.1  MAFIP  AC011043.1  AL592183.1  \\\n",
       "0            0.0     0.0  ...         0.0    0.0         0.0         0.0   \n",
       "1            0.0     0.0  ...         0.0    0.0         0.0         0.0   \n",
       "2            0.0     0.0  ...         0.0    0.0         0.0         0.0   \n",
       "3            0.0     0.0  ...         0.0    0.0         0.0         1.0   \n",
       "4            0.0     0.0  ...         0.0    0.0         0.0         0.0   \n",
       "...          ...     ...  ...         ...    ...         ...         ...   \n",
       "6075         0.0     0.0  ...         0.0    0.0         0.0         0.0   \n",
       "6076         0.0     0.0  ...         0.0    0.0         0.0         0.0   \n",
       "6077         0.0     0.0  ...         0.0    0.0         0.0         0.0   \n",
       "6078         0.0     0.0  ...         0.0    0.0         0.0         0.0   \n",
       "6079         0.0     0.0  ...         0.0    0.0         0.0         0.0   \n",
       "\n",
       "      AC007325.4  AL354822.1  AC004556.1  AC233755.2  AC233755.1  AC240274.1  \n",
       "0            0.0         0.0         0.0         0.0         0.0         1.0  \n",
       "1            0.0         2.0         1.0         0.0         0.0         1.0  \n",
       "2            0.0         1.0         0.0         0.0         0.0         0.0  \n",
       "3            0.0         0.0         0.0         0.0         0.0         1.0  \n",
       "4            0.0         0.0         4.0         0.0         0.0         0.0  \n",
       "...          ...         ...         ...         ...         ...         ...  \n",
       "6075         0.0         0.0         0.0         0.0         0.0         0.0  \n",
       "6076         0.0         0.0         0.0         0.0         0.0         0.0  \n",
       "6077         0.0         0.0         1.0         0.0         0.0         0.0  \n",
       "6078         0.0         0.0         0.0         0.0         0.0         0.0  \n",
       "6079         0.0         0.0         0.0         0.0         0.0         0.0  \n",
       "\n",
       "[6080 rows x 20606 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Just get the raw data, that's all I want."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NK_CD56++           320\n",
       "T4_RO+_SELL+        320\n",
       "T4_CM               320\n",
       "T4_RA+_SELLint      320\n",
       "T4_RA+_SELL+        320\n",
       "T_CD10+             320\n",
       "T_gd                320\n",
       "T4_RO+_Act          320\n",
       "T8_HOBIT+HELIOS+    320\n",
       "T8_TEMRA            320\n",
       "MAIT                320\n",
       "NK_CD16+            320\n",
       "T8_RA+_SELL+        320\n",
       "T4_Treg_Act         320\n",
       "T4_Treg_Resting     320\n",
       "T8_CM               320\n",
       "T8_EM               320\n",
       "cDC1                320\n",
       "cDC2                320\n",
       "Name: CT, dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts['CT'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['MAIT', 'NK_CD16+', 'NK_CD56++', 'T4_CM', 'T4_RA+_SELL+',\n",
       "       'T4_RA+_SELLint', 'T4_RO+_Act', 'T4_RO+_SELL+', 'T4_Treg_Act',\n",
       "       'T4_Treg_Resting', 'T8_CM', 'T8_EM', 'T8_HOBIT+HELIOS+',\n",
       "       'T8_RA+_SELL+', 'T8_TEMRA', 'T_CD10+', 'T_gd', 'cDC1', 'cDC2'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(counts['CT'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts.isna().sum().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create 5 separate sets of files, one for each stimulation, that has counts for both controls and the stim conditions, and in the appropriate format for easy reading into DESeq2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "99683bf0f54d4acbad92e7a3a8be8faf",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/5 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for cond in tqdm(['A','B','G','P','R']):\n",
    "    # extract out only the cond conditions and controls\n",
    "    df = counts[(counts['COND'] == cond) | (counts['COND'] == 'C')]\n",
    "    \n",
    "    # make a new index that encompasses the entirety of the sample name\n",
    "    df.index = ['-'.join([i,j,k]) for i,j,k in zip(df['CT'], df['COND'], df['FID'])]\n",
    "    \n",
    "\n",
    "    # With the bulk data, I had to remove and sum over duplicate genes\n",
    "    # but because this is pseuobulked from an adata object, I already ran \n",
    "    # var_names_make_unique, so I think I'm good (I also confirmed in a separate cell)\n",
    "    \n",
    "    # extract out only the genes, then rotate because that's what DESeq2 expects\n",
    "    cts = df.iloc[:,3:].T\n",
    "    \n",
    "    \n",
    "    # get new columns, and then make a separate dfs for the coldata\n",
    "    coldata_columns = df.columns[:3]\n",
    "    coldata = pd.DataFrame(data=df[coldata_columns].values, index=cts.columns, columns=coldata_columns)\n",
    "    \n",
    "    \n",
    "    # next line was required, was getting the following error from DESeq2:\n",
    "    # every gene contains at least one zero, cannot compute log geometric means\n",
    "    # this was the suggested fix:\n",
    "    cts = cts.loc[cts.sum(1) > 0,:] + 1\n",
    "\n",
    "    #export to csv\n",
    "    cts.to_csv(prefix_mrna + 'vals/de/all/ct3/input/%s_cts.csv' % cond)\n",
    "    coldata.to_csv(prefix_mrna + 'vals/de/all/ct3/input/%s_col.csv' % cond)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ADTs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = pd.read_pickle(prefix_adts + 'pkls/pseudobulk_ct3.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CT</th>\n",
       "      <th>COND</th>\n",
       "      <th>FID</th>\n",
       "      <th>CD103|ITGAE</th>\n",
       "      <th>CD137|TNFRSF9</th>\n",
       "      <th>CD274|CD274</th>\n",
       "      <th>CD11b|ITGAM</th>\n",
       "      <th>CD39|ENTPD1</th>\n",
       "      <th>CD197|CCR7</th>\n",
       "      <th>CD20|MS4A1</th>\n",
       "      <th>...</th>\n",
       "      <th>CD49a|ITGA1</th>\n",
       "      <th>CD133|PROM</th>\n",
       "      <th>CD357|TNFRSF18</th>\n",
       "      <th>CD270|TNFRSF14</th>\n",
       "      <th>IL-21R|IL21R</th>\n",
       "      <th>B7-H4|VTCN1</th>\n",
       "      <th>CD26|DPP4</th>\n",
       "      <th>CD155|PVR</th>\n",
       "      <th>CD30|TNFRSF8</th>\n",
       "      <th>CD49E|ITGA5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MAIT</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>20.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>513.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>98.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MAIT</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>30.0</td>\n",
       "      <td>376.0</td>\n",
       "      <td>203.0</td>\n",
       "      <td>145.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>257.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>...</td>\n",
       "      <td>69.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>281.0</td>\n",
       "      <td>362.0</td>\n",
       "      <td>455.0</td>\n",
       "      <td>4388.0</td>\n",
       "      <td>207.0</td>\n",
       "      <td>145.0</td>\n",
       "      <td>415.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MAIT</td>\n",
       "      <td>A</td>\n",
       "      <td>10</td>\n",
       "      <td>24.0</td>\n",
       "      <td>275.0</td>\n",
       "      <td>117.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>148.0</td>\n",
       "      <td>158.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>...</td>\n",
       "      <td>43.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>110.0</td>\n",
       "      <td>201.0</td>\n",
       "      <td>290.0</td>\n",
       "      <td>336.0</td>\n",
       "      <td>2133.0</td>\n",
       "      <td>168.0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>120.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MAIT</td>\n",
       "      <td>A</td>\n",
       "      <td>11</td>\n",
       "      <td>20.0</td>\n",
       "      <td>306.0</td>\n",
       "      <td>133.0</td>\n",
       "      <td>82.0</td>\n",
       "      <td>162.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>...</td>\n",
       "      <td>52.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>109.0</td>\n",
       "      <td>249.0</td>\n",
       "      <td>260.0</td>\n",
       "      <td>317.0</td>\n",
       "      <td>2520.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>206.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MAIT</td>\n",
       "      <td>A</td>\n",
       "      <td>12</td>\n",
       "      <td>8.0</td>\n",
       "      <td>176.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>101.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>...</td>\n",
       "      <td>31.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>141.0</td>\n",
       "      <td>228.0</td>\n",
       "      <td>1178.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>119.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6075</th>\n",
       "      <td>cDC2</td>\n",
       "      <td>R</td>\n",
       "      <td>62</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6076</th>\n",
       "      <td>cDC2</td>\n",
       "      <td>R</td>\n",
       "      <td>63</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6077</th>\n",
       "      <td>cDC2</td>\n",
       "      <td>R</td>\n",
       "      <td>7</td>\n",
       "      <td>17.0</td>\n",
       "      <td>497.0</td>\n",
       "      <td>495.0</td>\n",
       "      <td>113.0</td>\n",
       "      <td>127.0</td>\n",
       "      <td>163.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>...</td>\n",
       "      <td>107.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>175.0</td>\n",
       "      <td>124.0</td>\n",
       "      <td>178.0</td>\n",
       "      <td>600.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>954.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6078</th>\n",
       "      <td>cDC2</td>\n",
       "      <td>R</td>\n",
       "      <td>8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>14.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>99.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6079</th>\n",
       "      <td>cDC2</td>\n",
       "      <td>R</td>\n",
       "      <td>9</td>\n",
       "      <td>28.0</td>\n",
       "      <td>343.0</td>\n",
       "      <td>447.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>108.0</td>\n",
       "      <td>169.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>...</td>\n",
       "      <td>63.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>136.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>211.0</td>\n",
       "      <td>163.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>474.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6080 rows × 99 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        CT COND FID  CD103|ITGAE  CD137|TNFRSF9  CD274|CD274  CD11b|ITGAM  \\\n",
       "0     MAIT    A   0         11.0           59.0         23.0         27.0   \n",
       "1     MAIT    A   1         30.0          376.0        203.0        145.0   \n",
       "2     MAIT    A  10         24.0          275.0        117.0         73.0   \n",
       "3     MAIT    A  11         20.0          306.0        133.0         82.0   \n",
       "4     MAIT    A  12          8.0          176.0         77.0         81.0   \n",
       "...    ...  ...  ..          ...            ...          ...          ...   \n",
       "6075  cDC2    R  62          0.0            0.0          0.0          0.0   \n",
       "6076  cDC2    R  63          0.0            0.0          0.0          0.0   \n",
       "6077  cDC2    R   7         17.0          497.0        495.0        113.0   \n",
       "6078  cDC2    R   8          0.0           96.0         85.0          7.0   \n",
       "6079  cDC2    R   9         28.0          343.0        447.0         79.0   \n",
       "\n",
       "      CD39|ENTPD1  CD197|CCR7  CD20|MS4A1  ...  CD49a|ITGA1  CD133|PROM  \\\n",
       "0            33.0        38.0         5.0  ...         20.0        11.0   \n",
       "1           181.0       257.0        52.0  ...         69.0        54.0   \n",
       "2           148.0       158.0        25.0  ...         43.0        48.0   \n",
       "3           162.0       156.0        25.0  ...         52.0        47.0   \n",
       "4           101.0        99.0        16.0  ...         31.0        25.0   \n",
       "...           ...         ...         ...  ...          ...         ...   \n",
       "6075          0.0         0.0         0.0  ...          0.0         0.0   \n",
       "6076          0.0         0.0         0.0  ...          0.0         0.0   \n",
       "6077        127.0       163.0        14.0  ...        107.0        27.0   \n",
       "6078         19.0        49.0         5.0  ...         14.0         5.0   \n",
       "6079        108.0       169.0        15.0  ...         63.0        23.0   \n",
       "\n",
       "      CD357|TNFRSF18  CD270|TNFRSF14  IL-21R|IL21R  B7-H4|VTCN1  CD26|DPP4  \\\n",
       "0               26.0            44.0          54.0         78.0      513.0   \n",
       "1              150.0           281.0         362.0        455.0     4388.0   \n",
       "2              110.0           201.0         290.0        336.0     2133.0   \n",
       "3              109.0           249.0         260.0        317.0     2520.0   \n",
       "4               62.0           137.0         141.0        228.0     1178.0   \n",
       "...              ...             ...           ...          ...        ...   \n",
       "6075             0.0             0.0           0.0          0.0        0.0   \n",
       "6076             0.0             0.0           0.0          0.0        0.0   \n",
       "6077            81.0           175.0         124.0        178.0      600.0   \n",
       "6078            25.0            34.0          36.0         41.0       55.0   \n",
       "6079            59.0           136.0         146.0        211.0      163.0   \n",
       "\n",
       "      CD155|PVR  CD30|TNFRSF8  CD49E|ITGA5  \n",
       "0          24.0          14.0         98.0  \n",
       "1         207.0         145.0        415.0  \n",
       "2         168.0         106.0        120.0  \n",
       "3         147.0          78.0        206.0  \n",
       "4          95.0          55.0        119.0  \n",
       "...         ...           ...          ...  \n",
       "6075        0.0           0.0          0.0  \n",
       "6076        0.0           0.0          0.0  \n",
       "6077       53.0          61.0        954.0  \n",
       "6078       17.0          18.0         99.0  \n",
       "6079       75.0          73.0        474.0  \n",
       "\n",
       "[6080 rows x 99 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Just get the raw data, that's all I want."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NK_CD56++           320\n",
       "T4_RO+_SELL+        320\n",
       "T4_CM               320\n",
       "T4_RA+_SELLint      320\n",
       "T4_RA+_SELL+        320\n",
       "T_CD10+             320\n",
       "T_gd                320\n",
       "T4_RO+_Act          320\n",
       "T8_HOBIT+HELIOS+    320\n",
       "T8_TEMRA            320\n",
       "MAIT                320\n",
       "NK_CD16+            320\n",
       "T8_RA+_SELL+        320\n",
       "T4_Treg_Act         320\n",
       "T4_Treg_Resting     320\n",
       "T8_CM               320\n",
       "T8_EM               320\n",
       "cDC1                320\n",
       "cDC2                320\n",
       "Name: CT, dtype: int64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts['CT'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['MAIT', 'NK_CD16+', 'NK_CD56++', 'T4_CM', 'T4_RA+_SELL+',\n",
       "       'T4_RA+_SELLint', 'T4_RO+_Act', 'T4_RO+_SELL+', 'T4_Treg_Act',\n",
       "       'T4_Treg_Resting', 'T8_CM', 'T8_EM', 'T8_HOBIT+HELIOS+',\n",
       "       'T8_RA+_SELL+', 'T8_TEMRA', 'T_CD10+', 'T_gd', 'cDC1', 'cDC2'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(counts['CT'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts.isna().sum().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create 5 separate sets of files, one for each stimulation, that has counts for both controls and the stim conditions, and in the appropriate format for easy reading into DESeq2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "21cd5b2b281d4270989de199fc59e84a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/5 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for cond in tqdm(['A','B','G','P','R']):\n",
    "    # extract out only the cond conditions and controls\n",
    "    df = counts[(counts['COND'] == cond) | (counts['COND'] == 'C')]\n",
    "    \n",
    "    # make a new index that encompasses the entirety of the sample name\n",
    "    df.index = ['-'.join([i,j,k]) for i,j,k in zip(df['CT'], df['COND'], df['FID'])]\n",
    "    \n",
    "\n",
    "    # With the bulk data, I had to remove and sum over duplicate genes\n",
    "    # but because this is pseuobulked from an adata object, I already ran \n",
    "    # var_names_make_unique, so I think I'm good (also confirmed in a separate cell)\n",
    "    \n",
    "    # extract out only the genes, then rotate because that's what DESeq2 expects\n",
    "    cts = df.iloc[:,3:].T\n",
    "    \n",
    "    \n",
    "    # get new columns, and then make a separate dfs for the coldata\n",
    "    coldata_columns = df.columns[:3]\n",
    "    coldata = pd.DataFrame(data=df[coldata_columns].values, index=cts.columns, columns=coldata_columns)\n",
    "    \n",
    "#     if cond == 'P':\n",
    "#         cts = cts.loc[:,~np.any(np.stack([cts.columns.str.contains('Mono_C-C'), cts.columns.str.contains('Mono_NC-C')]), axis=0)]\n",
    "#         coldata = coldata.loc[~np.any(np.stack([coldata.index.str.contains('Mono_C-C'), coldata.index.str.contains('Mono_NC-C')]), axis=0),:]\n",
    "    \n",
    "    # next line was required, was getting the following error from DESeq2:\n",
    "    # every gene contains at least one zero, cannot compute log geometric means\n",
    "    # this was the suggested fix:\n",
    "    cts = cts.loc[cts.sum(1) > 0,:] + 1\n",
    "\n",
    "    #export to csv\n",
    "    cts.to_csv(prefix_adts + 'vals/de/all/ct3/input/%s_cts.csv' % cond)\n",
    "    coldata.to_csv(prefix_adts + 'vals/de/all/ct3/input/%s_col.csv' % cond)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# IFNs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## mRNA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ct2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = pd.read_pickle(prefix_mrna + 'pkls/pseudobulk_ct2.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CT</th>\n",
       "      <th>COND</th>\n",
       "      <th>FID</th>\n",
       "      <th>AL627309.1</th>\n",
       "      <th>AL627309.3</th>\n",
       "      <th>AL669831.5</th>\n",
       "      <th>LINC00115</th>\n",
       "      <th>FAM41C</th>\n",
       "      <th>AL645608.3</th>\n",
       "      <th>SAMD11</th>\n",
       "      <th>...</th>\n",
       "      <th>AC145212.1</th>\n",
       "      <th>MAFIP</th>\n",
       "      <th>AC011043.1</th>\n",
       "      <th>AL592183.1</th>\n",
       "      <th>AC007325.4</th>\n",
       "      <th>AL354822.1</th>\n",
       "      <th>AC004556.1</th>\n",
       "      <th>AC233755.2</th>\n",
       "      <th>AC233755.1</th>\n",
       "      <th>AC240274.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T4_Naive</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T4_Naive</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T4_Naive</td>\n",
       "      <td>A</td>\n",
       "      <td>10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T4_Naive</td>\n",
       "      <td>A</td>\n",
       "      <td>11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T4_Naive</td>\n",
       "      <td>A</td>\n",
       "      <td>12</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>M_cDC</td>\n",
       "      <td>R</td>\n",
       "      <td>62</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>M_cDC</td>\n",
       "      <td>R</td>\n",
       "      <td>63</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>M_cDC</td>\n",
       "      <td>R</td>\n",
       "      <td>7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>M_cDC</td>\n",
       "      <td>R</td>\n",
       "      <td>8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>M_cDC</td>\n",
       "      <td>R</td>\n",
       "      <td>9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4800 rows × 20606 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           CT COND FID  AL627309.1  AL627309.3  AL669831.5  LINC00115  FAM41C  \\\n",
       "0    T4_Naive    A   0         0.0         0.0         1.0        2.0     3.0   \n",
       "1    T4_Naive    A   1         0.0         0.0         6.0        2.0     3.0   \n",
       "2    T4_Naive    A  10         0.0         1.0        16.0        3.0     4.0   \n",
       "3    T4_Naive    A  11         0.0         0.0         0.0        0.0     4.0   \n",
       "4    T4_Naive    A  12         0.0         0.0         3.0        2.0     2.0   \n",
       "..        ...  ...  ..         ...         ...         ...        ...     ...   \n",
       "315     M_cDC    R  62         1.0         0.0         0.0        0.0     1.0   \n",
       "316     M_cDC    R  63         0.0         0.0         0.0        0.0     0.0   \n",
       "317     M_cDC    R   7         1.0         0.0         1.0        1.0     1.0   \n",
       "318     M_cDC    R   8         0.0         0.0         3.0        6.0     1.0   \n",
       "319     M_cDC    R   9         0.0         0.0         4.0        7.0     4.0   \n",
       "\n",
       "     AL645608.3  SAMD11  ...  AC145212.1  MAFIP  AC011043.1  AL592183.1  \\\n",
       "0           0.0     0.0  ...         0.0    0.0         0.0         5.0   \n",
       "1           0.0     0.0  ...         0.0    2.0         0.0         0.0   \n",
       "2           0.0     0.0  ...         1.0    0.0         1.0         4.0   \n",
       "3           1.0     0.0  ...         0.0    0.0         0.0         1.0   \n",
       "4           0.0     0.0  ...         0.0    1.0         0.0         4.0   \n",
       "..          ...     ...  ...         ...    ...         ...         ...   \n",
       "315         0.0     0.0  ...         0.0    0.0         0.0         0.0   \n",
       "316         0.0     0.0  ...         0.0    0.0         0.0         0.0   \n",
       "317         0.0     0.0  ...         0.0    1.0         0.0         1.0   \n",
       "318         0.0     0.0  ...         0.0    0.0         0.0         3.0   \n",
       "319         0.0     0.0  ...         0.0    0.0         0.0         2.0   \n",
       "\n",
       "     AC007325.4  AL354822.1  AC004556.1  AC233755.2  AC233755.1  AC240274.1  \n",
       "0           0.0         1.0         0.0         0.0         0.0         4.0  \n",
       "1           0.0         3.0         0.0         0.0         0.0         1.0  \n",
       "2           2.0         4.0         0.0         0.0         0.0         3.0  \n",
       "3           2.0         2.0         1.0         0.0         0.0         4.0  \n",
       "4           0.0         2.0        55.0         0.0         0.0         5.0  \n",
       "..          ...         ...         ...         ...         ...         ...  \n",
       "315         0.0         0.0         1.0         0.0         1.0         4.0  \n",
       "316         0.0         0.0         0.0         0.0         0.0         0.0  \n",
       "317         0.0         1.0         9.0         0.0         0.0         0.0  \n",
       "318         1.0         1.0         0.0         0.0         0.0         1.0  \n",
       "319         4.0         5.0         4.0         0.0         0.0         5.0  \n",
       "\n",
       "[4800 rows x 20606 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Just get the raw data, that's all I want."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "T8_Naive    384\n",
       "B_Mem       384\n",
       "NK          384\n",
       "pDC         384\n",
       "B_Naive     384\n",
       "HSC         384\n",
       "T4_EM       384\n",
       "M_cDC       384\n",
       "T_Tox       384\n",
       "T4_Naive    384\n",
       "ncM         320\n",
       "cM          320\n",
       "cDC         320\n",
       "Name: CT, dtype: int64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts['CT'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['B_Mem', 'B_Naive', 'HSC', 'M_cDC', 'NK', 'T4_EM', 'T4_Naive',\n",
       "       'T8_Naive', 'T_Tox', 'cDC', 'cM', 'ncM', 'pDC'], dtype=object)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(counts['CT'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts.isna().sum().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2051096f8f78442e8c0d8f82626f78f8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for cond in tqdm(['B']):\n",
    "    # extract out only the cond conditions and controls\n",
    "    df = counts[(counts['COND'] == cond) | (counts['COND'] == 'G')]\n",
    "    \n",
    "    # make a new index that encompasses the entirety of the sample name\n",
    "    df.index = ['-'.join([i,j,k]) for i,j,k in zip(df['CT'], df['COND'], df['FID'])]\n",
    "    \n",
    "\n",
    "    # With the bulk data, I had to remove and sum over duplicate genes\n",
    "    # but because this is pseuobulked from an adata object, I already ran \n",
    "    # var_names_make_unique, so I think I'm good (also confirmed in a separate cell)\n",
    "    \n",
    "    # extract out only the genes, then rotate because that's what DESeq2 expects\n",
    "    cts = df.iloc[:,3:].T\n",
    "    \n",
    "    \n",
    "    # get new columns, and then make a separate dfs for the coldata\n",
    "    coldata_columns = df.columns[:3]\n",
    "    coldata = pd.DataFrame(data=df[coldata_columns].values, index=cts.columns, columns=coldata_columns)\n",
    "    \n",
    "#     if cond == 'P':\n",
    "#         cts = cts.loc[:,~np.any(np.stack([cts.columns.str.contains('Mono_C-C'), cts.columns.str.contains('Mono_NC-C')]), axis=0)]\n",
    "#         coldata = coldata.loc[~np.any(np.stack([coldata.index.str.contains('Mono_C-C'), coldata.index.str.contains('Mono_NC-C')]), axis=0),:]\n",
    "    \n",
    "    # next line was required, was getting the following error from DESeq2:\n",
    "    # every gene contains at least one zero, cannot compute log geometric means\n",
    "    # this was the suggested fix:\n",
    "    cts = cts.loc[cts.sum(1) > 0,:] + 1\n",
    "\n",
    "    #export to csv\n",
    "    cts.to_csv(prefix_mrna + 'vals/de/IFNs/ct2/input/B-G_cts.csv')\n",
    "    coldata.to_csv(prefix_mrna + 'vals/de/IFNs/ct2/input/B-G_col.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2179917644cc4385b713e675dd0c249a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for cond in tqdm(['G']):\n",
    "    # extract out only the cond conditions and controls\n",
    "    df = counts[(counts['COND'] == cond) | (counts['COND'] == 'B')]\n",
    "    \n",
    "    # make a new index that encompasses the entirety of the sample name\n",
    "    df.index = ['-'.join([i,j,k]) for i,j,k in zip(df['CT'], df['COND'], df['FID'])]\n",
    "    \n",
    "\n",
    "    # With the bulk data, I had to remove and sum over duplicate genes\n",
    "    # but because this is pseuobulked from an adata object, I already ran \n",
    "    # var_names_make_unique, so I think I'm good (also confirmed in a separate cell)\n",
    "    \n",
    "    # extract out only the genes, then rotate because that's what DESeq2 expects\n",
    "    cts = df.iloc[:,3:].T\n",
    "    \n",
    "    \n",
    "    # get new columns, and then make a separate dfs for the coldata\n",
    "    coldata_columns = df.columns[:3]\n",
    "    coldata = pd.DataFrame(data=df[coldata_columns].values, index=cts.columns, columns=coldata_columns)\n",
    "    \n",
    "#     if cond == 'P':\n",
    "#         cts = cts.loc[:,~np.any(np.stack([cts.columns.str.contains('Mono_C-C'), cts.columns.str.contains('Mono_NC-C')]), axis=0)]\n",
    "#         coldata = coldata.loc[~np.any(np.stack([coldata.index.str.contains('Mono_C-C'), coldata.index.str.contains('Mono_NC-C')]), axis=0),:]\n",
    "    \n",
    "    # next line was required, was getting the following error from DESeq2:\n",
    "    # every gene contains at least one zero, cannot compute log geometric means\n",
    "    # this was the suggested fix:\n",
    "    cts = cts.loc[cts.sum(1) > 0,:] + 1\n",
    "\n",
    "    #export to csv\n",
    "    cts.to_csv(prefix_mrna + 'vals/de/IFNs/ct2/input/G-B_cts.csv')\n",
    "    coldata.to_csv(prefix_mrna + 'vals/de/IFNs/ct2/input/G-B_col.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ct3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = pd.read_pickle(prefix_mrna + 'pkls/pseudobulk_ct3.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CT</th>\n",
       "      <th>COND</th>\n",
       "      <th>FID</th>\n",
       "      <th>AL627309.1</th>\n",
       "      <th>AL627309.3</th>\n",
       "      <th>AL669831.5</th>\n",
       "      <th>LINC00115</th>\n",
       "      <th>FAM41C</th>\n",
       "      <th>AL645608.3</th>\n",
       "      <th>SAMD11</th>\n",
       "      <th>...</th>\n",
       "      <th>AC145212.1</th>\n",
       "      <th>MAFIP</th>\n",
       "      <th>AC011043.1</th>\n",
       "      <th>AL592183.1</th>\n",
       "      <th>AC007325.4</th>\n",
       "      <th>AL354822.1</th>\n",
       "      <th>AC004556.1</th>\n",
       "      <th>AC233755.2</th>\n",
       "      <th>AC233755.1</th>\n",
       "      <th>AC240274.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MAIT</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MAIT</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MAIT</td>\n",
       "      <td>A</td>\n",
       "      <td>10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MAIT</td>\n",
       "      <td>A</td>\n",
       "      <td>11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MAIT</td>\n",
       "      <td>A</td>\n",
       "      <td>12</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6075</th>\n",
       "      <td>cDC2</td>\n",
       "      <td>R</td>\n",
       "      <td>62</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6076</th>\n",
       "      <td>cDC2</td>\n",
       "      <td>R</td>\n",
       "      <td>63</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6077</th>\n",
       "      <td>cDC2</td>\n",
       "      <td>R</td>\n",
       "      <td>7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6078</th>\n",
       "      <td>cDC2</td>\n",
       "      <td>R</td>\n",
       "      <td>8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6079</th>\n",
       "      <td>cDC2</td>\n",
       "      <td>R</td>\n",
       "      <td>9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6080 rows × 20606 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        CT COND FID  AL627309.1  AL627309.3  AL669831.5  LINC00115  FAM41C  \\\n",
       "0     MAIT    A   0         0.0         0.0         0.0        0.0     0.0   \n",
       "1     MAIT    A   1         0.0         0.0         1.0        3.0     0.0   \n",
       "2     MAIT    A  10         0.0         0.0         1.0        0.0     0.0   \n",
       "3     MAIT    A  11         0.0         0.0         0.0        0.0     0.0   \n",
       "4     MAIT    A  12         0.0         0.0         0.0        0.0     0.0   \n",
       "...    ...  ...  ..         ...         ...         ...        ...     ...   \n",
       "6075  cDC2    R  62         0.0         0.0         0.0        0.0     0.0   \n",
       "6076  cDC2    R  63         0.0         0.0         0.0        0.0     0.0   \n",
       "6077  cDC2    R   7         0.0         0.0         0.0        0.0     0.0   \n",
       "6078  cDC2    R   8         0.0         0.0         3.0        0.0     0.0   \n",
       "6079  cDC2    R   9         0.0         0.0         0.0        1.0     0.0   \n",
       "\n",
       "      AL645608.3  SAMD11  ...  AC145212.1  MAFIP  AC011043.1  AL592183.1  \\\n",
       "0            0.0     0.0  ...         0.0    0.0         0.0         0.0   \n",
       "1            0.0     0.0  ...         0.0    0.0         0.0         0.0   \n",
       "2            0.0     0.0  ...         0.0    0.0         0.0         0.0   \n",
       "3            0.0     0.0  ...         0.0    0.0         0.0         1.0   \n",
       "4            0.0     0.0  ...         0.0    0.0         0.0         0.0   \n",
       "...          ...     ...  ...         ...    ...         ...         ...   \n",
       "6075         0.0     0.0  ...         0.0    0.0         0.0         0.0   \n",
       "6076         0.0     0.0  ...         0.0    0.0         0.0         0.0   \n",
       "6077         0.0     0.0  ...         0.0    0.0         0.0         0.0   \n",
       "6078         0.0     0.0  ...         0.0    0.0         0.0         0.0   \n",
       "6079         0.0     0.0  ...         0.0    0.0         0.0         0.0   \n",
       "\n",
       "      AC007325.4  AL354822.1  AC004556.1  AC233755.2  AC233755.1  AC240274.1  \n",
       "0            0.0         0.0         0.0         0.0         0.0         1.0  \n",
       "1            0.0         2.0         1.0         0.0         0.0         1.0  \n",
       "2            0.0         1.0         0.0         0.0         0.0         0.0  \n",
       "3            0.0         0.0         0.0         0.0         0.0         1.0  \n",
       "4            0.0         0.0         4.0         0.0         0.0         0.0  \n",
       "...          ...         ...         ...         ...         ...         ...  \n",
       "6075         0.0         0.0         0.0         0.0         0.0         0.0  \n",
       "6076         0.0         0.0         0.0         0.0         0.0         0.0  \n",
       "6077         0.0         0.0         1.0         0.0         0.0         0.0  \n",
       "6078         0.0         0.0         0.0         0.0         0.0         0.0  \n",
       "6079         0.0         0.0         0.0         0.0         0.0         0.0  \n",
       "\n",
       "[6080 rows x 20606 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Just get the raw data, that's all I want."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NK_CD56++           320\n",
       "T4_RO+_SELL+        320\n",
       "T4_CM               320\n",
       "T4_RA+_SELLint      320\n",
       "T4_RA+_SELL+        320\n",
       "T_CD10+             320\n",
       "T_gd                320\n",
       "T4_RO+_Act          320\n",
       "T8_HOBIT+HELIOS+    320\n",
       "T8_TEMRA            320\n",
       "MAIT                320\n",
       "NK_CD16+            320\n",
       "T8_RA+_SELL+        320\n",
       "T4_Treg_Act         320\n",
       "T4_Treg_Resting     320\n",
       "T8_CM               320\n",
       "T8_EM               320\n",
       "cDC1                320\n",
       "cDC2                320\n",
       "Name: CT, dtype: int64"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts['CT'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['MAIT', 'NK_CD16+', 'NK_CD56++', 'T4_CM', 'T4_RA+_SELL+',\n",
       "       'T4_RA+_SELLint', 'T4_RO+_Act', 'T4_RO+_SELL+', 'T4_Treg_Act',\n",
       "       'T4_Treg_Resting', 'T8_CM', 'T8_EM', 'T8_HOBIT+HELIOS+',\n",
       "       'T8_RA+_SELL+', 'T8_TEMRA', 'T_CD10+', 'T_gd', 'cDC1', 'cDC2'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(counts['CT'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts.isna().sum().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6525ebbbe4a448e19dd3ce17def437b4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for cond in tqdm(['B']):\n",
    "    # extract out only the cond conditions and controls\n",
    "    df = counts[(counts['COND'] == cond) | (counts['COND'] == 'G')]\n",
    "    \n",
    "    # make a new index that encompasses the entirety of the sample name\n",
    "    df.index = ['-'.join([i,j,k]) for i,j,k in zip(df['CT'], df['COND'], df['FID'])]\n",
    "    \n",
    "\n",
    "    # With the bulk data, I had to remove and sum over duplicate genes\n",
    "    # but because this is pseuobulked from an adata object, I already ran \n",
    "    # var_names_make_unique, so I think I'm good (also confirmed in a separate cell)\n",
    "    \n",
    "    # extract out only the genes, then rotate because that's what DESeq2 expects\n",
    "    cts = df.iloc[:,3:].T\n",
    "    \n",
    "    \n",
    "    # get new columns, and then make a separate dfs for the coldata\n",
    "    coldata_columns = df.columns[:3]\n",
    "    coldata = pd.DataFrame(data=df[coldata_columns].values, index=cts.columns, columns=coldata_columns)\n",
    "    \n",
    "#     if cond == 'P':\n",
    "#         cts = cts.loc[:,~np.any(np.stack([cts.columns.str.contains('Mono_C-C'), cts.columns.str.contains('Mono_NC-C')]), axis=0)]\n",
    "#         coldata = coldata.loc[~np.any(np.stack([coldata.index.str.contains('Mono_C-C'), coldata.index.str.contains('Mono_NC-C')]), axis=0),:]\n",
    "    \n",
    "    # next line was required, was getting the following error from DESeq2:\n",
    "    # every gene contains at least one zero, cannot compute log geometric means\n",
    "    # this was the suggested fix:\n",
    "    cts = cts.loc[cts.sum(1) > 0,:] + 1\n",
    "\n",
    "    #export to csv\n",
    "    cts.to_csv(prefix_mrna + 'vals/de/IFNs/ct3/input/B-G_cts.csv')\n",
    "    coldata.to_csv(prefix_mrna + 'vals/de/IFNs/ct3/input/B-G_col.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b0efcee17bac4dcaaf75e89ddc378195",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for cond in tqdm(['G']):\n",
    "    # extract out only the cond conditions and controls\n",
    "    df = counts[(counts['COND'] == cond) | (counts['COND'] == 'B')]\n",
    "    \n",
    "    # make a new index that encompasses the entirety of the sample name\n",
    "    df.index = ['-'.join([i,j,k]) for i,j,k in zip(df['CT'], df['COND'], df['FID'])]\n",
    "    \n",
    "\n",
    "    # With the bulk data, I had to remove and sum over duplicate genes\n",
    "    # but because this is pseuobulked from an adata object, I already ran \n",
    "    # var_names_make_unique, so I think I'm good (also confirmed in a separate cell)\n",
    "    \n",
    "    # extract out only the genes, then rotate because that's what DESeq2 expects\n",
    "    cts = df.iloc[:,3:].T\n",
    "    \n",
    "    \n",
    "    # get new columns, and then make a separate dfs for the coldata\n",
    "    coldata_columns = df.columns[:3]\n",
    "    coldata = pd.DataFrame(data=df[coldata_columns].values, index=cts.columns, columns=coldata_columns)\n",
    "    \n",
    "#     if cond == 'P':\n",
    "#         cts = cts.loc[:,~np.any(np.stack([cts.columns.str.contains('Mono_C-C'), cts.columns.str.contains('Mono_NC-C')]), axis=0)]\n",
    "#         coldata = coldata.loc[~np.any(np.stack([coldata.index.str.contains('Mono_C-C'), coldata.index.str.contains('Mono_NC-C')]), axis=0),:]\n",
    "    \n",
    "    # next line was required, was getting the following error from DESeq2:\n",
    "    # every gene contains at least one zero, cannot compute log geometric means\n",
    "    # this was the suggested fix:\n",
    "    cts = cts.loc[cts.sum(1) > 0,:] + 1\n",
    "\n",
    "    #export to csv\n",
    "    cts.to_csv(prefix_mrna + 'vals/de/IFNs/ct3/input/G-B_cts.csv')\n",
    "    coldata.to_csv(prefix_mrna + 'vals/de/IFNs/ct3/input/G-B_col.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ADTs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ct2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = pd.read_pickle(prefix_adts + 'pkls/pseudobulk_ct2.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CT</th>\n",
       "      <th>COND</th>\n",
       "      <th>FID</th>\n",
       "      <th>CD103|ITGAE</th>\n",
       "      <th>CD137|TNFRSF9</th>\n",
       "      <th>CD274|CD274</th>\n",
       "      <th>CD11b|ITGAM</th>\n",
       "      <th>CD39|ENTPD1</th>\n",
       "      <th>CD197|CCR7</th>\n",
       "      <th>CD20|MS4A1</th>\n",
       "      <th>...</th>\n",
       "      <th>CD49a|ITGA1</th>\n",
       "      <th>CD133|PROM</th>\n",
       "      <th>CD357|TNFRSF18</th>\n",
       "      <th>CD270|TNFRSF14</th>\n",
       "      <th>IL-21R|IL21R</th>\n",
       "      <th>B7-H4|VTCN1</th>\n",
       "      <th>CD26|DPP4</th>\n",
       "      <th>CD155|PVR</th>\n",
       "      <th>CD30|TNFRSF8</th>\n",
       "      <th>CD49E|ITGA5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T4_Naive</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>118.0</td>\n",
       "      <td>1590.0</td>\n",
       "      <td>987.0</td>\n",
       "      <td>805.0</td>\n",
       "      <td>1028.0</td>\n",
       "      <td>1907.0</td>\n",
       "      <td>236.0</td>\n",
       "      <td>...</td>\n",
       "      <td>320.0</td>\n",
       "      <td>313.0</td>\n",
       "      <td>790.0</td>\n",
       "      <td>1637.0</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>2463.0</td>\n",
       "      <td>6362.0</td>\n",
       "      <td>1131.0</td>\n",
       "      <td>668.0</td>\n",
       "      <td>1248.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T4_Naive</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>199.0</td>\n",
       "      <td>1950.0</td>\n",
       "      <td>1438.0</td>\n",
       "      <td>774.0</td>\n",
       "      <td>1329.0</td>\n",
       "      <td>2990.0</td>\n",
       "      <td>379.0</td>\n",
       "      <td>...</td>\n",
       "      <td>436.0</td>\n",
       "      <td>450.0</td>\n",
       "      <td>1017.0</td>\n",
       "      <td>1981.0</td>\n",
       "      <td>2645.0</td>\n",
       "      <td>3334.0</td>\n",
       "      <td>5773.0</td>\n",
       "      <td>1577.0</td>\n",
       "      <td>1020.0</td>\n",
       "      <td>1603.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T4_Naive</td>\n",
       "      <td>A</td>\n",
       "      <td>10</td>\n",
       "      <td>501.0</td>\n",
       "      <td>5193.0</td>\n",
       "      <td>3877.0</td>\n",
       "      <td>2439.0</td>\n",
       "      <td>3893.0</td>\n",
       "      <td>6985.0</td>\n",
       "      <td>1121.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1254.0</td>\n",
       "      <td>1210.0</td>\n",
       "      <td>2989.0</td>\n",
       "      <td>5679.0</td>\n",
       "      <td>7913.0</td>\n",
       "      <td>9536.0</td>\n",
       "      <td>19177.0</td>\n",
       "      <td>4672.0</td>\n",
       "      <td>2866.0</td>\n",
       "      <td>3612.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T4_Naive</td>\n",
       "      <td>A</td>\n",
       "      <td>11</td>\n",
       "      <td>139.0</td>\n",
       "      <td>1732.0</td>\n",
       "      <td>1207.0</td>\n",
       "      <td>674.0</td>\n",
       "      <td>1178.0</td>\n",
       "      <td>2347.0</td>\n",
       "      <td>358.0</td>\n",
       "      <td>...</td>\n",
       "      <td>369.0</td>\n",
       "      <td>384.0</td>\n",
       "      <td>955.0</td>\n",
       "      <td>1940.0</td>\n",
       "      <td>2613.0</td>\n",
       "      <td>3226.0</td>\n",
       "      <td>5598.0</td>\n",
       "      <td>1538.0</td>\n",
       "      <td>879.0</td>\n",
       "      <td>1332.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T4_Naive</td>\n",
       "      <td>A</td>\n",
       "      <td>12</td>\n",
       "      <td>203.0</td>\n",
       "      <td>2777.0</td>\n",
       "      <td>1857.0</td>\n",
       "      <td>1052.0</td>\n",
       "      <td>1891.0</td>\n",
       "      <td>3274.0</td>\n",
       "      <td>495.0</td>\n",
       "      <td>...</td>\n",
       "      <td>629.0</td>\n",
       "      <td>599.0</td>\n",
       "      <td>1515.0</td>\n",
       "      <td>2977.0</td>\n",
       "      <td>3999.0</td>\n",
       "      <td>5079.0</td>\n",
       "      <td>7580.0</td>\n",
       "      <td>2146.0</td>\n",
       "      <td>1405.0</td>\n",
       "      <td>2027.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>M_cDC</td>\n",
       "      <td>R</td>\n",
       "      <td>62</td>\n",
       "      <td>134.0</td>\n",
       "      <td>3320.0</td>\n",
       "      <td>2444.0</td>\n",
       "      <td>6729.0</td>\n",
       "      <td>1420.0</td>\n",
       "      <td>1220.0</td>\n",
       "      <td>268.0</td>\n",
       "      <td>...</td>\n",
       "      <td>602.0</td>\n",
       "      <td>324.0</td>\n",
       "      <td>909.0</td>\n",
       "      <td>2307.0</td>\n",
       "      <td>1939.0</td>\n",
       "      <td>2473.0</td>\n",
       "      <td>1786.0</td>\n",
       "      <td>1147.0</td>\n",
       "      <td>763.0</td>\n",
       "      <td>11072.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>M_cDC</td>\n",
       "      <td>R</td>\n",
       "      <td>63</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>M_cDC</td>\n",
       "      <td>R</td>\n",
       "      <td>7</td>\n",
       "      <td>353.0</td>\n",
       "      <td>6702.0</td>\n",
       "      <td>17348.0</td>\n",
       "      <td>12877.0</td>\n",
       "      <td>2107.0</td>\n",
       "      <td>1628.0</td>\n",
       "      <td>290.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1105.0</td>\n",
       "      <td>342.0</td>\n",
       "      <td>1073.0</td>\n",
       "      <td>3566.0</td>\n",
       "      <td>2115.0</td>\n",
       "      <td>2496.0</td>\n",
       "      <td>3555.0</td>\n",
       "      <td>1199.0</td>\n",
       "      <td>1030.0</td>\n",
       "      <td>16470.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>M_cDC</td>\n",
       "      <td>R</td>\n",
       "      <td>8</td>\n",
       "      <td>148.0</td>\n",
       "      <td>2819.0</td>\n",
       "      <td>6210.0</td>\n",
       "      <td>3463.0</td>\n",
       "      <td>1342.0</td>\n",
       "      <td>1460.0</td>\n",
       "      <td>220.0</td>\n",
       "      <td>...</td>\n",
       "      <td>489.0</td>\n",
       "      <td>311.0</td>\n",
       "      <td>861.0</td>\n",
       "      <td>1944.0</td>\n",
       "      <td>1829.0</td>\n",
       "      <td>2745.0</td>\n",
       "      <td>1844.0</td>\n",
       "      <td>1073.0</td>\n",
       "      <td>779.0</td>\n",
       "      <td>8548.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>M_cDC</td>\n",
       "      <td>R</td>\n",
       "      <td>9</td>\n",
       "      <td>678.0</td>\n",
       "      <td>15417.0</td>\n",
       "      <td>37167.0</td>\n",
       "      <td>24157.0</td>\n",
       "      <td>5873.0</td>\n",
       "      <td>5586.0</td>\n",
       "      <td>1144.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2596.0</td>\n",
       "      <td>1252.0</td>\n",
       "      <td>3610.0</td>\n",
       "      <td>9344.0</td>\n",
       "      <td>7920.0</td>\n",
       "      <td>9767.0</td>\n",
       "      <td>7297.0</td>\n",
       "      <td>4626.0</td>\n",
       "      <td>3459.0</td>\n",
       "      <td>46142.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4800 rows × 99 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           CT COND FID  CD103|ITGAE  CD137|TNFRSF9  CD274|CD274  CD11b|ITGAM  \\\n",
       "0    T4_Naive    A   0        118.0         1590.0        987.0        805.0   \n",
       "1    T4_Naive    A   1        199.0         1950.0       1438.0        774.0   \n",
       "2    T4_Naive    A  10        501.0         5193.0       3877.0       2439.0   \n",
       "3    T4_Naive    A  11        139.0         1732.0       1207.0        674.0   \n",
       "4    T4_Naive    A  12        203.0         2777.0       1857.0       1052.0   \n",
       "..        ...  ...  ..          ...            ...          ...          ...   \n",
       "315     M_cDC    R  62        134.0         3320.0       2444.0       6729.0   \n",
       "316     M_cDC    R  63          0.0            0.0          0.0          0.0   \n",
       "317     M_cDC    R   7        353.0         6702.0      17348.0      12877.0   \n",
       "318     M_cDC    R   8        148.0         2819.0       6210.0       3463.0   \n",
       "319     M_cDC    R   9        678.0        15417.0      37167.0      24157.0   \n",
       "\n",
       "     CD39|ENTPD1  CD197|CCR7  CD20|MS4A1  ...  CD49a|ITGA1  CD133|PROM  \\\n",
       "0         1028.0      1907.0       236.0  ...        320.0       313.0   \n",
       "1         1329.0      2990.0       379.0  ...        436.0       450.0   \n",
       "2         3893.0      6985.0      1121.0  ...       1254.0      1210.0   \n",
       "3         1178.0      2347.0       358.0  ...        369.0       384.0   \n",
       "4         1891.0      3274.0       495.0  ...        629.0       599.0   \n",
       "..           ...         ...         ...  ...          ...         ...   \n",
       "315       1420.0      1220.0       268.0  ...        602.0       324.0   \n",
       "316          0.0         0.0         0.0  ...          0.0         0.0   \n",
       "317       2107.0      1628.0       290.0  ...       1105.0       342.0   \n",
       "318       1342.0      1460.0       220.0  ...        489.0       311.0   \n",
       "319       5873.0      5586.0      1144.0  ...       2596.0      1252.0   \n",
       "\n",
       "     CD357|TNFRSF18  CD270|TNFRSF14  IL-21R|IL21R  B7-H4|VTCN1  CD26|DPP4  \\\n",
       "0             790.0          1637.0        2019.0       2463.0     6362.0   \n",
       "1            1017.0          1981.0        2645.0       3334.0     5773.0   \n",
       "2            2989.0          5679.0        7913.0       9536.0    19177.0   \n",
       "3             955.0          1940.0        2613.0       3226.0     5598.0   \n",
       "4            1515.0          2977.0        3999.0       5079.0     7580.0   \n",
       "..              ...             ...           ...          ...        ...   \n",
       "315           909.0          2307.0        1939.0       2473.0     1786.0   \n",
       "316             0.0             0.0           0.0          0.0        0.0   \n",
       "317          1073.0          3566.0        2115.0       2496.0     3555.0   \n",
       "318           861.0          1944.0        1829.0       2745.0     1844.0   \n",
       "319          3610.0          9344.0        7920.0       9767.0     7297.0   \n",
       "\n",
       "     CD155|PVR  CD30|TNFRSF8  CD49E|ITGA5  \n",
       "0       1131.0         668.0       1248.0  \n",
       "1       1577.0        1020.0       1603.0  \n",
       "2       4672.0        2866.0       3612.0  \n",
       "3       1538.0         879.0       1332.0  \n",
       "4       2146.0        1405.0       2027.0  \n",
       "..         ...           ...          ...  \n",
       "315     1147.0         763.0      11072.0  \n",
       "316        0.0           0.0          0.0  \n",
       "317     1199.0        1030.0      16470.0  \n",
       "318     1073.0         779.0       8548.0  \n",
       "319     4626.0        3459.0      46142.0  \n",
       "\n",
       "[4800 rows x 99 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Just get the raw data, that's all I want."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "T8_Naive    384\n",
       "B_Mem       384\n",
       "NK          384\n",
       "pDC         384\n",
       "B_Naive     384\n",
       "HSC         384\n",
       "T4_EM       384\n",
       "M_cDC       384\n",
       "T_Tox       384\n",
       "T4_Naive    384\n",
       "ncM         320\n",
       "cM          320\n",
       "cDC         320\n",
       "Name: CT, dtype: int64"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts['CT'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['B_Mem', 'B_Naive', 'HSC', 'M_cDC', 'NK', 'T4_EM', 'T4_Naive',\n",
       "       'T8_Naive', 'T_Tox', 'cDC', 'cM', 'ncM', 'pDC'], dtype=object)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(counts['CT'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts.isna().sum().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create 5 separate sets of files, one for each stimulation, that has counts for both controls and the stim conditions, and in the appropriate format for easy reading into DESeq2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "53aa8f89c6304ce0aae59874defef881",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for cond in tqdm(['B']):\n",
    "    # extract out only the cond conditions and controls\n",
    "    df = counts[(counts['COND'] == cond) | (counts['COND'] == 'G')]\n",
    "    \n",
    "    # make a new index that encompasses the entirety of the sample name\n",
    "    df.index = ['-'.join([i,j,k]) for i,j,k in zip(df['CT'], df['COND'], df['FID'])]\n",
    "    \n",
    "\n",
    "    # With the bulk data, I had to remove and sum over duplicate genes\n",
    "    # but because this is pseuobulked from an adata object, I already ran \n",
    "    # var_names_make_unique, so I think I'm good (also confirmed in a separate cell)\n",
    "    \n",
    "    # extract out only the genes, then rotate because that's what DESeq2 expects\n",
    "    cts = df.iloc[:,3:].T\n",
    "    \n",
    "    \n",
    "    # get new columns, and then make a separate dfs for the coldata\n",
    "    coldata_columns = df.columns[:3]\n",
    "    coldata = pd.DataFrame(data=df[coldata_columns].values, index=cts.columns, columns=coldata_columns)\n",
    "    \n",
    "#     if cond == 'P':\n",
    "#         cts = cts.loc[:,~np.any(np.stack([cts.columns.str.contains('Mono_C-C'), cts.columns.str.contains('Mono_NC-C')]), axis=0)]\n",
    "#         coldata = coldata.loc[~np.any(np.stack([coldata.index.str.contains('Mono_C-C'), coldata.index.str.contains('Mono_NC-C')]), axis=0),:]\n",
    "    \n",
    "    # next line was required, was getting the following error from DESeq2:\n",
    "    # every gene contains at least one zero, cannot compute log geometric means\n",
    "    # this was the suggested fix:\n",
    "    cts = cts.loc[cts.sum(1) > 0,:] + 1\n",
    "\n",
    "    #export to csv\n",
    "    cts.to_csv(prefix_adts + 'vals/de/IFNs/ct2/input/B-G_cts.csv')\n",
    "    coldata.to_csv(prefix_adts + 'vals/de/IFNs/ct2/input/B-G_col.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3dd9468b5d2a4f168e41c0c20b2e494e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for cond in tqdm(['G']):\n",
    "    # extract out only the cond conditions and controls\n",
    "    df = counts[(counts['COND'] == cond) | (counts['COND'] == 'B')]\n",
    "    \n",
    "    # make a new index that encompasses the entirety of the sample name\n",
    "    df.index = ['-'.join([i,j,k]) for i,j,k in zip(df['CT'], df['COND'], df['FID'])]\n",
    "    \n",
    "\n",
    "    # With the bulk data, I had to remove and sum over duplicate genes\n",
    "    # but because this is pseuobulked from an adata object, I already ran \n",
    "    # var_names_make_unique, so I think I'm good (also confirmed in a separate cell)\n",
    "    \n",
    "    # extract out only the genes, then rotate because that's what DESeq2 expects\n",
    "    cts = df.iloc[:,3:].T\n",
    "    \n",
    "    \n",
    "    # get new columns, and then make a separate dfs for the coldata\n",
    "    coldata_columns = df.columns[:3]\n",
    "    coldata = pd.DataFrame(data=df[coldata_columns].values, index=cts.columns, columns=coldata_columns)\n",
    "    \n",
    "#     if cond == 'P':\n",
    "#         cts = cts.loc[:,~np.any(np.stack([cts.columns.str.contains('Mono_C-C'), cts.columns.str.contains('Mono_NC-C')]), axis=0)]\n",
    "#         coldata = coldata.loc[~np.any(np.stack([coldata.index.str.contains('Mono_C-C'), coldata.index.str.contains('Mono_NC-C')]), axis=0),:]\n",
    "    \n",
    "    # next line was required, was getting the following error from DESeq2:\n",
    "    # every gene contains at least one zero, cannot compute log geometric means\n",
    "    # this was the suggested fix:\n",
    "    cts = cts.loc[cts.sum(1) > 0,:] + 1\n",
    "\n",
    "    #export to csv\n",
    "    cts.to_csv(prefix_adts + 'vals/de/IFNs/ct2/input/G-B_cts.csv')\n",
    "    coldata.to_csv(prefix_adts + 'vals/de/IFNs/ct2/input/G-B_col.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ct3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = pd.read_pickle(prefix_adts + 'pkls/pseudobulk_ct3.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CT</th>\n",
       "      <th>COND</th>\n",
       "      <th>FID</th>\n",
       "      <th>CD103|ITGAE</th>\n",
       "      <th>CD137|TNFRSF9</th>\n",
       "      <th>CD274|CD274</th>\n",
       "      <th>CD11b|ITGAM</th>\n",
       "      <th>CD39|ENTPD1</th>\n",
       "      <th>CD197|CCR7</th>\n",
       "      <th>CD20|MS4A1</th>\n",
       "      <th>...</th>\n",
       "      <th>CD49a|ITGA1</th>\n",
       "      <th>CD133|PROM</th>\n",
       "      <th>CD357|TNFRSF18</th>\n",
       "      <th>CD270|TNFRSF14</th>\n",
       "      <th>IL-21R|IL21R</th>\n",
       "      <th>B7-H4|VTCN1</th>\n",
       "      <th>CD26|DPP4</th>\n",
       "      <th>CD155|PVR</th>\n",
       "      <th>CD30|TNFRSF8</th>\n",
       "      <th>CD49E|ITGA5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MAIT</td>\n",
       "      <td>A</td>\n",
       "      <td>0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>20.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>513.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>98.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MAIT</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>30.0</td>\n",
       "      <td>376.0</td>\n",
       "      <td>203.0</td>\n",
       "      <td>145.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>257.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>...</td>\n",
       "      <td>69.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>281.0</td>\n",
       "      <td>362.0</td>\n",
       "      <td>455.0</td>\n",
       "      <td>4388.0</td>\n",
       "      <td>207.0</td>\n",
       "      <td>145.0</td>\n",
       "      <td>415.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MAIT</td>\n",
       "      <td>A</td>\n",
       "      <td>10</td>\n",
       "      <td>24.0</td>\n",
       "      <td>275.0</td>\n",
       "      <td>117.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>148.0</td>\n",
       "      <td>158.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>...</td>\n",
       "      <td>43.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>110.0</td>\n",
       "      <td>201.0</td>\n",
       "      <td>290.0</td>\n",
       "      <td>336.0</td>\n",
       "      <td>2133.0</td>\n",
       "      <td>168.0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>120.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MAIT</td>\n",
       "      <td>A</td>\n",
       "      <td>11</td>\n",
       "      <td>20.0</td>\n",
       "      <td>306.0</td>\n",
       "      <td>133.0</td>\n",
       "      <td>82.0</td>\n",
       "      <td>162.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>...</td>\n",
       "      <td>52.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>109.0</td>\n",
       "      <td>249.0</td>\n",
       "      <td>260.0</td>\n",
       "      <td>317.0</td>\n",
       "      <td>2520.0</td>\n",
       "      <td>147.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>206.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MAIT</td>\n",
       "      <td>A</td>\n",
       "      <td>12</td>\n",
       "      <td>8.0</td>\n",
       "      <td>176.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>101.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>...</td>\n",
       "      <td>31.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>141.0</td>\n",
       "      <td>228.0</td>\n",
       "      <td>1178.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>119.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6075</th>\n",
       "      <td>cDC2</td>\n",
       "      <td>R</td>\n",
       "      <td>62</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6076</th>\n",
       "      <td>cDC2</td>\n",
       "      <td>R</td>\n",
       "      <td>63</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6077</th>\n",
       "      <td>cDC2</td>\n",
       "      <td>R</td>\n",
       "      <td>7</td>\n",
       "      <td>17.0</td>\n",
       "      <td>497.0</td>\n",
       "      <td>495.0</td>\n",
       "      <td>113.0</td>\n",
       "      <td>127.0</td>\n",
       "      <td>163.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>...</td>\n",
       "      <td>107.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>175.0</td>\n",
       "      <td>124.0</td>\n",
       "      <td>178.0</td>\n",
       "      <td>600.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>954.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6078</th>\n",
       "      <td>cDC2</td>\n",
       "      <td>R</td>\n",
       "      <td>8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>14.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>99.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6079</th>\n",
       "      <td>cDC2</td>\n",
       "      <td>R</td>\n",
       "      <td>9</td>\n",
       "      <td>28.0</td>\n",
       "      <td>343.0</td>\n",
       "      <td>447.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>108.0</td>\n",
       "      <td>169.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>...</td>\n",
       "      <td>63.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>136.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>211.0</td>\n",
       "      <td>163.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>474.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6080 rows × 99 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        CT COND FID  CD103|ITGAE  CD137|TNFRSF9  CD274|CD274  CD11b|ITGAM  \\\n",
       "0     MAIT    A   0         11.0           59.0         23.0         27.0   \n",
       "1     MAIT    A   1         30.0          376.0        203.0        145.0   \n",
       "2     MAIT    A  10         24.0          275.0        117.0         73.0   \n",
       "3     MAIT    A  11         20.0          306.0        133.0         82.0   \n",
       "4     MAIT    A  12          8.0          176.0         77.0         81.0   \n",
       "...    ...  ...  ..          ...            ...          ...          ...   \n",
       "6075  cDC2    R  62          0.0            0.0          0.0          0.0   \n",
       "6076  cDC2    R  63          0.0            0.0          0.0          0.0   \n",
       "6077  cDC2    R   7         17.0          497.0        495.0        113.0   \n",
       "6078  cDC2    R   8          0.0           96.0         85.0          7.0   \n",
       "6079  cDC2    R   9         28.0          343.0        447.0         79.0   \n",
       "\n",
       "      CD39|ENTPD1  CD197|CCR7  CD20|MS4A1  ...  CD49a|ITGA1  CD133|PROM  \\\n",
       "0            33.0        38.0         5.0  ...         20.0        11.0   \n",
       "1           181.0       257.0        52.0  ...         69.0        54.0   \n",
       "2           148.0       158.0        25.0  ...         43.0        48.0   \n",
       "3           162.0       156.0        25.0  ...         52.0        47.0   \n",
       "4           101.0        99.0        16.0  ...         31.0        25.0   \n",
       "...           ...         ...         ...  ...          ...         ...   \n",
       "6075          0.0         0.0         0.0  ...          0.0         0.0   \n",
       "6076          0.0         0.0         0.0  ...          0.0         0.0   \n",
       "6077        127.0       163.0        14.0  ...        107.0        27.0   \n",
       "6078         19.0        49.0         5.0  ...         14.0         5.0   \n",
       "6079        108.0       169.0        15.0  ...         63.0        23.0   \n",
       "\n",
       "      CD357|TNFRSF18  CD270|TNFRSF14  IL-21R|IL21R  B7-H4|VTCN1  CD26|DPP4  \\\n",
       "0               26.0            44.0          54.0         78.0      513.0   \n",
       "1              150.0           281.0         362.0        455.0     4388.0   \n",
       "2              110.0           201.0         290.0        336.0     2133.0   \n",
       "3              109.0           249.0         260.0        317.0     2520.0   \n",
       "4               62.0           137.0         141.0        228.0     1178.0   \n",
       "...              ...             ...           ...          ...        ...   \n",
       "6075             0.0             0.0           0.0          0.0        0.0   \n",
       "6076             0.0             0.0           0.0          0.0        0.0   \n",
       "6077            81.0           175.0         124.0        178.0      600.0   \n",
       "6078            25.0            34.0          36.0         41.0       55.0   \n",
       "6079            59.0           136.0         146.0        211.0      163.0   \n",
       "\n",
       "      CD155|PVR  CD30|TNFRSF8  CD49E|ITGA5  \n",
       "0          24.0          14.0         98.0  \n",
       "1         207.0         145.0        415.0  \n",
       "2         168.0         106.0        120.0  \n",
       "3         147.0          78.0        206.0  \n",
       "4          95.0          55.0        119.0  \n",
       "...         ...           ...          ...  \n",
       "6075        0.0           0.0          0.0  \n",
       "6076        0.0           0.0          0.0  \n",
       "6077       53.0          61.0        954.0  \n",
       "6078       17.0          18.0         99.0  \n",
       "6079       75.0          73.0        474.0  \n",
       "\n",
       "[6080 rows x 99 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Just get the raw data, that's all I want."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NK_CD56++           320\n",
       "T4_RO+_SELL+        320\n",
       "T4_CM               320\n",
       "T4_RA+_SELLint      320\n",
       "T4_RA+_SELL+        320\n",
       "T_CD10+             320\n",
       "T_gd                320\n",
       "T4_RO+_Act          320\n",
       "T8_HOBIT+HELIOS+    320\n",
       "T8_TEMRA            320\n",
       "MAIT                320\n",
       "NK_CD16+            320\n",
       "T8_RA+_SELL+        320\n",
       "T4_Treg_Act         320\n",
       "T4_Treg_Resting     320\n",
       "T8_CM               320\n",
       "T8_EM               320\n",
       "cDC1                320\n",
       "cDC2                320\n",
       "Name: CT, dtype: int64"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts['CT'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['MAIT', 'NK_CD16+', 'NK_CD56++', 'T4_CM', 'T4_RA+_SELL+',\n",
       "       'T4_RA+_SELLint', 'T4_RO+_Act', 'T4_RO+_SELL+', 'T4_Treg_Act',\n",
       "       'T4_Treg_Resting', 'T8_CM', 'T8_EM', 'T8_HOBIT+HELIOS+',\n",
       "       'T8_RA+_SELL+', 'T8_TEMRA', 'T_CD10+', 'T_gd', 'cDC1', 'cDC2'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(counts['CT'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts.isna().sum().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create 5 separate sets of files, one for each stimulation, that has counts for both controls and the stim conditions, and in the appropriate format for easy reading into DESeq2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "aaf55c9945d14719b53d81a5a9b730ee",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for cond in tqdm(['B']):\n",
    "    # extract out only the cond conditions and controls\n",
    "    df = counts[(counts['COND'] == cond) | (counts['COND'] == 'G')]\n",
    "    \n",
    "    # make a new index that encompasses the entirety of the sample name\n",
    "    df.index = ['-'.join([i,j,k]) for i,j,k in zip(df['CT'], df['COND'], df['FID'])]\n",
    "    \n",
    "\n",
    "    # With the bulk data, I had to remove and sum over duplicate genes\n",
    "    # but because this is pseuobulked from an adata object, I already ran \n",
    "    # var_names_make_unique, so I think I'm good (also confirmed in a separate cell)\n",
    "    \n",
    "    # extract out only the genes, then rotate because that's what DESeq2 expects\n",
    "    cts = df.iloc[:,3:].T\n",
    "    \n",
    "    \n",
    "    # get new columns, and then make a separate dfs for the coldata\n",
    "    coldata_columns = df.columns[:3]\n",
    "    coldata = pd.DataFrame(data=df[coldata_columns].values, index=cts.columns, columns=coldata_columns)\n",
    "    \n",
    "#     if cond == 'P':\n",
    "#         cts = cts.loc[:,~np.any(np.stack([cts.columns.str.contains('Mono_C-C'), cts.columns.str.contains('Mono_NC-C')]), axis=0)]\n",
    "#         coldata = coldata.loc[~np.any(np.stack([coldata.index.str.contains('Mono_C-C'), coldata.index.str.contains('Mono_NC-C')]), axis=0),:]\n",
    "    \n",
    "    # next line was required, was getting the following error from DESeq2:\n",
    "    # every gene contains at least one zero, cannot compute log geometric means\n",
    "    # this was the suggested fix:\n",
    "    cts = cts.loc[cts.sum(1) > 0,:] + 1\n",
    "\n",
    "    #export to csv\n",
    "    cts.to_csv(prefix_adts + 'vals/de/IFNs/ct3/input/B-G_cts.csv')\n",
    "    coldata.to_csv(prefix_adts + 'vals/de/IFNs/ct3/input/B-G_col.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e8ba38e56c264608b95f70bb945c8654",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for cond in tqdm(['G']):\n",
    "    # extract out only the cond conditions and controls\n",
    "    df = counts[(counts['COND'] == cond) | (counts['COND'] == 'B')]\n",
    "    \n",
    "    # make a new index that encompasses the entirety of the sample name\n",
    "    df.index = ['-'.join([i,j,k]) for i,j,k in zip(df['CT'], df['COND'], df['FID'])]\n",
    "    \n",
    "\n",
    "    # With the bulk data, I had to remove and sum over duplicate genes\n",
    "    # but because this is pseuobulked from an adata object, I already ran \n",
    "    # var_names_make_unique, so I think I'm good (also confirmed in a separate cell)\n",
    "    \n",
    "    # extract out only the genes, then rotate because that's what DESeq2 expects\n",
    "    cts = df.iloc[:,3:].T\n",
    "    \n",
    "    \n",
    "    # get new columns, and then make a separate dfs for the coldata\n",
    "    coldata_columns = df.columns[:3]\n",
    "    coldata = pd.DataFrame(data=df[coldata_columns].values, index=cts.columns, columns=coldata_columns)\n",
    "    \n",
    "#     if cond == 'P':\n",
    "#         cts = cts.loc[:,~np.any(np.stack([cts.columns.str.contains('Mono_C-C'), cts.columns.str.contains('Mono_NC-C')]), axis=0)]\n",
    "#         coldata = coldata.loc[~np.any(np.stack([coldata.index.str.contains('Mono_C-C'), coldata.index.str.contains('Mono_NC-C')]), axis=0),:]\n",
    "    \n",
    "    # next line was required, was getting the following error from DESeq2:\n",
    "    # every gene contains at least one zero, cannot compute log geometric means\n",
    "    # this was the suggested fix:\n",
    "    cts = cts.loc[cts.sum(1) > 0,:] + 1\n",
    "\n",
    "    #export to csv\n",
    "    cts.to_csv(prefix_adts + 'vals/de/IFNs/ct3/input/G-B_cts.csv')\n",
    "    coldata.to_csv(prefix_adts + 'vals/de/IFNs/ct3/input/G-B_col.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "clue",
   "language": "python",
   "name": "clue"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
